conans/model/build_info.py

```diff
@@ -1,224 +1,572 @@
+                     "include_paths", "src_paths", "lib_paths", "bin_paths", "build_paths", "res_paths",
+                     "defines", "cflags", "cxxflags", "cppflags", "sharedlinkflags", "exelinkflags"]:
+            attr_name = "cxxflags" if name == "cppflags" else name  # Backwards compatibility
+            result[name] = getattr(self, attr_name)
+        result["components"] = {}
+        for name, component in self.components.items():
+            result["components"][name] = component.as_dict()
+        result["configs"] = {}
+        for config, dep_cpp_info in self.configs.items():
+            result["configs"][config] = dep_cpp_info.as_dict()
+        return result
+
+
+class DepsCppInfo(object):
+    """ Build Information necessary to build a given conans. It contains the
+    flags, directories and options if its dependencies. The conans CONANFILE
+    should use these flags to pass them to the underlaying build system (Cmake, make),
+    so deps info is managed
     """
-    def __init__(self, root_folder):
-        super(CppInfo, self).__init__()
-        self.rootpath = root_folder  # the full path of the package in which the conans is found
-        self.includedirs.append(DEFAULT_INCLUDE)
-        self.libdirs.append(DEFAULT_LIB)
-        self.bindirs.append(DEFAULT_BIN)
-        self.resdirs.append(DEFAULT_RES)
-        self.builddirs.append("")
-        # public_deps is needed to accumulate list of deps for cmake targets
-        self.public_deps = []
+
+    def __init__(self):
+        self._dependencies = OrderedDict()
         self.configs = {}
 
     def __getattr__(self, config):
+        # If the configuration does not exist, return an empty list
+        if config not in self.configs:
+            # FIXME: This could create unintended empty configurations for those generators/libs
+            # that access unexisting configs with self.deps_cpp_info.whatever.includedirs
+            self.configs[config] = DepsCppInfo()
+        return self.configs[config]
 
-        def _get_cpp_info():
-            result = _CppInfo()
-            result.rootpath = self.rootpath
-            result.sysroot = self.sysroot
-            result.includedirs.append(DEFAULT_INCLUDE)
-            result.libdirs.append(DEFAULT_LIB)
-            result.bindirs.append(DEFAULT_BIN)
-            result.resdirs.append(DEFAULT_RES)
-            result.builddirs.append("")
-            return result
+    @property
+    def dependencies(self):
+        return self._dependencies.items()
+
+    @property
+    def deps(self):
+        return self._dependencies.keys()
 
-        return self.configs.setdefault(config, _get_cpp_info())
+    def __getitem__(self, item):
+        return self._dependencies[item]
 
+    def update(self, cpp_info, pkg_name):
+        assert isinstance(cpp_info, CppInfo)
+        self.update_dep_cpp_info(DepCppInfo(cpp_info), pkg_name)
 
-class _BaseDepsCppInfo(_CppInfo):
```

Comment:
This is mostly a problem of mixing interfaces. This class was used as the parent class of the object ``deps_cpp_info``, which had no-sense values such as `self.deps_cpp_info.version` or `self.deps_cpp_info.description`. Also the computation of the aggregated values is not done in the same way: `_BaseDepsInfo` (`DepCppInfo` in this PR) aggregates its values from Components (It is "a list of components"), while `DepsCppInfo` (Same name in this PR) aggregates its values from the list of dependencies (It is "a list of nodes/references/conanfiles").

We may argue that those classes to avoid some repetition, but I also think it is more clear to have them separated in two different classes.

Reply:
Probably yes, but please, take into account that too many changes in a PR, especially if they are not exactly intended for the main purpose of the issue, will stuck this to be released because the review will be harder and the possibilities of breaking behavior increase.